### PR TITLE
Search: Basic shard state

### PR DESCRIPTION
**Pull request by dranikpg**
Context: 

This is a preparation step before adding basic indices to core/search. Indices will be managed per shard, need to be constructed when the index is created with ft.create and need to be updated when new entries are added/deleted. Indices will store integer codes (compact & faster) instead of keys for documents. 

Goal:

* Spread out search_family parts into multiple files to make room for bigger future changes
* Support accessible per-shard state that will at first store only document key <-> integer id mapping. No updates

### CODE REVIEW

File: src/server/search/search_family.cc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  vector<vector<SerializedSearchDoc>> docs(shard_set->size());
+
+  cntx->transaction->ScheduleSingleHop([&](Transaction* t, EngineShard* es) {
+    if (auto* index = ShardSearchIndex::GetOnShard(index_name); index)
+      docs[es->shard_id()] = index->Search(t->GetOpArgs(es), &*search_algo);
+    else
+      index_not_found.store(true, memory_order_relaxed);
+    return OpStatus::OK;
+  });
+
```
Comment (dranikpg) [inline src/server/search/search_family.cc:98]: If we store indices per-shard, we can't answer in the coordinator fiber if it exists or not. The other solution is maintaining just a list of valid names on all threads

File: src/server/search/search_accessor.h [hunk 2]
```diff
@@ -0,0 +1,167 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (romange) [reply src/server/search/search_accessor.h:5]: how do you know it stays this way?

File: src/server/search/search_index.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "core/search/search.h"
+#include "server/common.h"
+
+namespace dfly {
+
+using SearchDocData = absl::flat_hash_map<std::string, std::string>;
```
Comment (romange) [inline src/server/search/search_index.h:17]: what does it map?

File: src/server/search/search_accessor.h [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+}  // namespace
+
+// Document accessors allow different types (json/hset) to be hidden
+// behind a document interface for quering fields and serializing.
+struct BaseAccessor : public search::DocumentAccessor {
+  using FieldConsumer = search::DocumentAccessor::FieldConsumer;
+
+  virtual SearchDocData Serialize() const = 0;
```
Comment (romange) [inline src/server/search/search_accessor.h:37]: remind what `Serialize` is used for?

File: src/server/search/search_family.cc [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  (*cntx)->SendOk();
+}
+
+void SearchFamily::FtSearch(CmdArgList args, ConnectionContext* cntx) {
+  string_view index_name = ArgS(args, 0);
+  string_view query_str = ArgS(args, 1);
+
+  optional<search::SearchAlgorithm> search_algo{};
+  try {
```
Comment (romange) [inline src/server/search/search_family.cc:80]: I avoid using c++ exceptions in general unless I do not have any options. 
In this case, seems you can just introduce an `Init` function.

File: src/server/search/search_family.cc [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    return;
+  }
+
+  // Because our coordinator thread may not have a shard, we can't check ahead if the index exists.
+  atomic<bool> index_not_found{false};
+  vector<vector<SerializedSearchDoc>> docs(shard_set->size());
+
+  cntx->transaction->ScheduleSingleHop([&](Transaction* t, EngineShard* es) {
+    if (auto* index = ShardSearchIndex::GetOnShard(index_name); index)
+      docs[es->shard_id()] = index->Search(t->GetOpArgs(es), &*search_algo);
```
Comment (romange) [inline src/server/search/search_family.cc:93]: we also won't need optional around `search_algo`

File: src/server/search/search_family.cc (hunk 1 above)
Comment (romange) [reply src/server/search/search_family.cc:98]: I think it's fine like this.

File: src/server/search/search_index.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+ private:
+  // Initialize index. Traverses all matching documents and assigns ids.
+  void Init(const OpArgs& op_args);
+
+ private:
+  std::shared_ptr<const SearchIndex> base_;
+  DocKeyIndex key_index_;
+
+  static thread_local absl::flat_hash_map<std::string, ShardSearchIndex> indices_;
```
[the remark this answers is not in the document]
Comment (romange) [reply src/server/search/search_index.h:68]: Yes, I think `ShardSearchIndex` should be there similarly to tiered.

File: src/server/search/search_index.cc [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+template <typename F>
+void TraverseAllMatching(const SearchIndex& index, const OpArgs& op_args, F&& f) {
+  auto& db_slice = op_args.shard->db_slice();
+  DCHECK(db_slice.IsDbValid(op_args.db_cntx.db_index));
+  auto [prime_table, _] = db_slice.GetTables(op_args.db_cntx.db_index);
+
+  string scratch;
+  auto cb = [&](PrimeTable::iterator it) {
+    // Check entry is hash
```
Comment (romange) [inline src/server/search/search_index.cc:23]: the comment is not in sync

File: src/server/search/search_index.h [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  using DocId = uint32_t;
+
+  // DocKeyIndex manages mapping document keys to ids and vice versa through a simple interface.
+  struct DocKeyIndex {
+    DocId Add(std::string_view key);
+    void Delete(std::string_view key);
+    std::string Get(DocId id);
+
+   private:
+    absl::flat_hash_map<std::string, DocId> ids;
```
Comment (romange) [inline src/server/search/search_index.h:43]: underscores at the end

File: src/server/search/search_accessor.h (hunk 4 above)
Comment (dranikpg) [reply src/server/search/search_accessor.h:37]: Convert underlying type to a map<string, string> to be sent as a reply, added a comment

**Comment by dranikpg** [comment]
Fixed comments. 

I also did some naming changes, because there will be around 5 depth levels and each one can be called index. Randomly adding words like shard or search doesn't make it more clear 🙂 

I suggest calling "logical indices" just document indices (those defined by ft.create SCHEMA PREFIX...). This indicates that they don't perform search, but just keep track of documents.

Shard prefix just means its per shard.

Now its like this:

DocIndex , ShardDocIndices -> ShardDocIndex -> [in pr2] -> FieldIndices -> BaseIndex -> e.g. NumericIndex

File: src/core/search/search.cc [hunk 10]
```diff
@@ -79,12 +79,17 @@ struct BasicSearch {
 SearchAlgorithm::SearchAlgorithm() = default;
 SearchAlgorithm::~SearchAlgorithm() = default;
 
-SearchAlgorithm::SearchAlgorithm(string_view query)
-    : query{make_unique<AstExpr>(ParseQuery(query))} {
+bool SearchAlgorithm::Init(string_view query) {
+  try {
+    query_ = make_unique<AstExpr>(ParseQuery(query));
```
Comment (romange) [inline src/core/search/search.cc:84]: make a comment explaining why do you try/catch.
catch the exception thrown by parser, ideally vlog it as well.

File: src/core/search/search_parser_test.cc [hunk 11]
```diff
@@ -36,19 +36,19 @@ class SearchParserTest : public ::testing::Test {
 
   void ParseExpr(const std::string& str) {
     raw_expr_ = str;
-    search_algo_.emplace(str);
+    search_algo_.Init(str);
```
Comment (romange) [inline src/core/search/search_parser_test.cc:39]: CHECK(...)

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]

**Pull request merged**